### PR TITLE
refine test connection logic

### PR DESCRIPTION
**Pull request by martinyunify**

### CODE REVIEW

File: backends/metad/client.go [hunk 1]
```diff
@@ -98,31 +84,48 @@ func NewMetadClient(backendNodes []string) (*Client, error) {
...
-		conn = c.connections.Value.(*Connection)
-		if conn == startConn {
-			return errors.New("Fail to connect any backend.")
+		conn := c.connections.Value.(*Connection)
+		startConn := conn
+
+		//iterate through connection
+		_, err = conn.makeMetaDataRequest("/")
+		for err != nil {
+			log.Error("Connection to [%s], error: [%s]", conn.url, err.Error())
```
Comment (chai2010) [inline backends/metad/client.go]: `log.Error("Connection to [%s], error: [%v]", conn.url, err)`

File: backends/metad/client.go [hunk 2]
```diff
@@ -98,31 +84,48 @@ func NewMetadClient(backendNodes []string) (*Client, error) {
...
+			c.connections = c.connections.Next()
+			conn = c.connections.Value.(*Connection)
+			if conn == startConn {
+				break
+			}
+			_, err = conn.makeMetaDataRequest("/")
```
Comment (chai2010) [inline backends/metad/client.go]: logic is complex

File: backends/metad/client.go [hunk 3]
```diff
@@ -98,31 +84,48 @@ func NewMetadClient(backendNodes []string) (*Client, error) {
...
 		}
-		err = conn.testConnection()
+		if err == nil {
+			//found available conn
+			if c.current != nil {
+				atomic.StoreUint32(&c.current.errTimes, 0)
+			}
+			c.current = conn
+			break
```
Comment (chai2010) [inline backends/metad/client.go:96]: err is nil means return ok ?

File: backends/metad/client.go [hunk 4]
```diff
@@ -98,31 +84,48 @@ func NewMetadClient(backendNodes []string) (*Client, error) {
...
+		}
+		time.Sleep(i)
 	}
-	if c.current != nil {
-		atomic.StoreUint32(&c.current.errTimes, 0)
+	if i < maxTime {
```
Comment (chai2010) [inline backends/metad/client.go]: better code style:

```go
if someFailed {
  return err
}

return nil // ok
```

File: backends/metad/client.go [hunk 5]
```diff
@@ -98,31 +84,48 @@ func NewMetadClient(backendNodes []string) (*Client, error) {
...
+		log.Info("Using Metad URL: " + c.current.url)
+		return nil
+	} else {
+		return fmt.Errorf("Fail to connect any backend.")
```
Comment (chai2010) [inline backends/metad/client.go]: ditto

File: backends/metad/client.go (hunk 2 above)
Comment (martinyunify) [reply backends/metad/client.go]: move the inner loop to testConnection function.

**Review by chai2010 — APPROVED** [review]

**Pull request merged**